### DISCUSSION TITLE
@types/datatables.net still necessary for DT v2?

### DISCUSSION THREAD

**Discussion by pfbristow** [l-lin/angular-datatables#1777] [General]
I'm a little late upgrading to datatables.net v2, and while I'm doing so, I'm noticing there are now TS type definitions: https://github.com/DataTables/DataTablesSrc/tree/master/types.  I'm not a very experienced TS / Angular developer, but these look complete, and it looks like they make @types/datatables.net obsolete for any project using the v2 API.

I tried uninstalling the @types/datatables.net package from my project, and after changing all my code to use the new types in v2 (much has changed and doesn't match the v1 API, though you can still use the v1 API, I think), the only errors I got from ng build were from angular-datatables:

```
Error: node_modules/angular-datatables/src/angular-datatables.directive.d.ts:31:25 - error TS2503: Cannot find namespace 'DataTables'.
31     dtInstance: Promise<DataTables.Api>;
                           ~~~~~~~~~~
Error: node_modules/angular-datatables/src/models/settings.d.ts:3:38 - error TS2503: Cannot find namespace 'DataTables'.
3 export interface ADTSettings extends DataTables.Settings {
                                       ~~~~~~~~~~
Error: node_modules/angular-datatables/src/models/settings.d.ts:6:37 - error TS2503: Cannot find namespace 'DataTables'.
6 export interface ADTColumns extends DataTables.ColumnSettings {
                                      ~~~~~~~~~~
```

So I re-installed @types/datatables.net, and then got errors due to conflicts in the APIs:

```
Error: features/claims/parties-involved-list.component.ts:50:44 - error TS2345: Argument of type '(dtInstance: Api) => void' is not assignable to parameter of type '(value: Api) => void | PromiseLike<void>'.
  Types of parameters 'dtInstance' and 'value' are incompatible.
    Type 'Api' is not assignable to type 'Api<any>'.

50     this.datatableElement!.dtInstance.then((dtInstance: Api) => {
                                              ~~~~~~~~~~~~~~~~~~~~~~


Error: node_modules/datatables.net/types/types.d.ts:242:9 - error TS2717: Subsequent property declarations must have the same type.  Property 'DataTable' must be of type '(opts?: Settings | undefined) => Api', but here has type 'JQueryDataTableApi'.

242         DataTable: JQueryDataTableApi;
            ~~~~~~~~~

  node_modules/@types/datatables.net/index.d.ts:7:5
    7     DataTable(opts?: DataTables.Settings): DataTables.Api;
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    'DataTable' was also declared here.


Error: node_modules/datatables.net/types/types.d.ts:250:9 - error TS2717: Subsequent property declarations must have the same type.  Property 'dataTable' must be of type 'StaticFunctions', but here has type 'JQueryDataTableJq'.

250         dataTable: JQueryDataTableJq;
            ~~~~~~~~~

  node_modules/@types/datatables.net/index.d.ts:8:5
    8     dataTable: DataTables.StaticFunctions;
          ~~~~~~~~~
    'dataTable' was also declared here.


Error: shared/base-datatable.component.ts:158:44 - error TS2345: Argument of type '(dtInstance: Api) => void' is not assignable to parameter of type '(value: Api) => void | PromiseLike<void>'.
  Types of parameters 'dtInstance' and 'value' are incompatible.
    Type 'Api' is missing the following properties from type 'Api<any>': caption, context, error, iterator, and 3 more.

158     this.datatableElement!.dtInstance.then((dtInstance: Api) => {
                                               ~~~~~~~~~~~~~~~~~~~~~~
```

I eventually realized I cannot reference the new v2 APIs direction from the DataTables type definitions, or else it causes build conflicts with angular-datatable's references to @types/datatables.net.

My first question is whether I'm doing anything wrong or misunderstanding that the new type definitions including in DataTables supercede @types/datatables.net?

If not, my second question is whether angular-datatables plan to migrate to the new API?

Apologies if any of these questions are silly or obvious.  I searched open & closed questions, commits, and issues.  I spend 95% of my time doing C# dev, so what little time I do TS & Angular I often feel a bit lost and n00b-like.  Thanks.

**Comment by shanmukhateja** [comment]
Hi, we currently don't support DataTables v2.

I am unable to find time to research this right now.

We need to maintain compatibility with both versions if possible.

A PR is always welcome :)

**Reply by pfbristow** [reply]
Ok thanks.  So far, I've gotten everything working except for search.  The API docs all seem to say that the old v1 API will be supported until v3 or otherwise a very long time, so that should be good for compatibility.  I do not see *anything* in the docs about the TypeScript definitions, which is odd since they're clearly in the source.

The only real problem so far as trying to figure out these type conflicts.  I couldn't find anything in angular-datatables that said it did or didn't support v2, and I couldn't find anything in datatables.net that said those TypeScript type definitions existed.  Hopefully anyone else struggling with the same thing will at least find this post.

Would you mind if I created an Issue for DataTables.net v2 support?  That might help increase visibility that you're looking for a PR and help others be aware that it's not yet supported.

**Comment by AllanJard** [comment]
Just a quick note from me to say that DaTables has shipped its own types file since 1.11 and the API has diverged from the `@types/datatables.net` package, which will indeed cause conflicts and issues. I would suggest that `@types/datatables.net` be removed, but if there are types from that package used in this, it might be a bit more involved than just `npm rm ...` :).

**edit** Worth noting that I also did a lot of renaming of interfaces around 1.13 (or 1.12 - can't recall off the top of my head!) to make them a little more sane. If 1.x compatibility is required, I'd suggest locking it to ^1.13. That said, DataTables 1.x is no longer supported by me (unless the customer has a maintenance package). I'd urge updates to 2.x.

**Reply by shanmukhateja** [reply]
Hey @pbarranis

Thanks for testing out the library and providing feedback. We usually don't get responses after a few days have passed on a ticket. Glad to know things still work.

So regarding type definition issues, I think I know what's happening. In recent versions of this library, we introduced interfaces that extend definitions defined by `@types/dataTables.net`  to support Angular specific features (pipes, template ref, etc.).
So when you remove `@types/datatables.net`, our library is unable to set the types for these custom interfaces (maybe because namespace `DataTables` is missing).

And in vice-versa, it freaks out because of conflicting type definitions (one is from built-in stuff in v2 and the other is the @types package), so it's a lose-lose situation for now.

I am currently not sure how to fix type conflicts without breaking v1 compatibility. Need to think it over.

Yes - please feel free to file a issue regarding v2 support. It will help track what APIs are broken at the least.  We could probably use checkboxes for tracking items.

**Reply by pfbristow** [reply]
Hi @AllanJard, will you be adding a page to the datatables.net manual about the Typescript type definitions?  I searched the site carefully for any mention of the types, and when I didn't find a page, I was very confused.

I'd also like to put in a request to DefinitelyTyped that the [@types/datatables.net](https://www.npmjs.com/package/@types/datatables.net?activeTab=versions) package be deprecated (to reduce confusion such as mine), but that's much clearer to do when there's a URL I can point to clearly saying the types are published directly from the package.

Thanks.

**Reply by AllanJard** [reply]
I've added a note to [my download builder](https://datatables.net/download/) when you select the `npm` option (i.e. yarn, etc as well) to state that Typescript files are included automatically.

I've also opened this discussion](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/69176) on Definitely Typed to see how I should remove / deprecate the old typing files.

**Reply by shanmukhateja** [reply]
Hello all, a quick progress update on this.

I have successfully been able to upgrade to v2 and include some developer experience fixes.

The unit tests are passing too which is a good sign.

I need to fix and test `ng add` support before I can raise PR.

> @AllanJard Can you take a look at types for fixed columns in NPM? I was not able to get v5 or v4 working with DT v2

Stay tuned!

**Reply by shanmukhateja** [reply]
As promised, I have upgraded the library to DT v2.

This branch will shift things around in the codebase (which isn't relevant to end-users) and finally brings in DT v2 support.

Feel free to build the PR branch locally and test things around. I (may have) missed few things. 

**PR:**
https://github.com/l-lin/angular-datatables/pull/1783

CC: @AllanJard @pbarranis

**Reply by AllanJard** [reply]
@shanmukhateja Awesome - nice one! I look forward to seeing it released :-).

**Comment by AllanJard** [comment]
Responsive 3.0.1 is on npm and should address the issues with typing that I'm aware of.
